### PR TITLE
ActiveRecord connection pool fixes

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

File: lib/dynflow/rails/configuration.rb [hunk 1]
```diff
@@ -100,8 +100,13 @@ def increase_db_pool_size?
...
       def calculate_db_pool_size(world)
-        self.db_pool_size || world.config.queues.values.inject(5) do |pool_size, pool_options|
-          pool_size += pool_options[:pool_size]
+        base_value = 5
+        if defined?(::Sidekiq)
+          Sidekiq.options[:concurrency] + base_value
+        else
+          self.db_pool_size || world.config.queues.values.inject(base_value) do |pool_size, pool_options|
+            pool_size += pool_options[:pool_size]
+          end
```
Comment (jlsherrill) [inline lib/dynflow/rails/configuration.rb:109]: if i'm reading this right, it means with sidekiq, we can no longer configure the db_size (independently of sidekiq's concurrency setting).  Do you see anyone ever needing to do that?

Comment (adamruzicka) [reply lib/dynflow/rails/configuration.rb:109]: That wasn't intentional, I'll put it back. It served us well in the past

File: lib/dynflow/executors.rb [hunk 2]
```diff
@@ -8,10 +8,11 @@ module Executors
     # we should wrap it with this method, and we can ensure here to do
     # necessary cleanup, such as cleaning ActiveRecord connections
     def self.run_user_code
-      clear_connections = defined?(::ActiveRecord) && ActiveRecord::Base.connected? && ActiveRecord::Base.connection.open_transactions.zero?
       yield
     ensure
-      ::ActiveRecord::Base.clear_active_connections! if clear_connections
+      if defined?(::ActiveRecord) && ActiveRecord::Base.connected? && ActiveRecord::Base.connection.open_transactions.zero?
```
Comment (ezr-ondrej) [inline lib/dynflow/executors.rb:13]: so now, we do not clear the connection if it's still active? It doesn't get cleared sometimes then?

**Comment by adamruzicka** [comment]
```
2020-04-14T12:12:08.228Z 586 TID-ou6cop396 WARN: ActiveRecord::ConnectionTimeoutError: could not obtain a connection from the pool within 5.000 seconds (waited 5.000 seconds); all pooled connections were in use
2020-04-14T12:12:08.228Z 586 TID-ou6cop396 WARN: /opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:199:in `block in wait_poll'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:190:in `loop'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:190:in `wait_poll'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:151:in `internal_poll'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:275:in `internal_poll'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:145:in `block in poll'
/opt/rh/rh-ruby25/root/usr/share/ruby/monitor.rb:226:in `mon_synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:155:in `synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:145:in `poll'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:797:in `acquire_connection'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:521:in `checkout'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:380:in `connection'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:1008:in `retrieve_connection'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_handling.rb:118:in `retrieve_connection'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_handling.rb:90:in `connection'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.3/lib/dynflow/executors.rb:11:in `run_user_code'
-----B<-----SNIP-----B<-----
```

`ActiveRecord::Base.connected?` apparently does something different than what I thought. According to the stacktrace `ActiveRecord::Base.connection` tries to obtain a *new* connection even if `ActiveRecord::Base.connected?` is true and that's where the issue comes from.

I have a patch in the works, need to run a few rounds of testing to rule out false positives.

File: lib/dynflow/executors.rb (hunk 2 above)
Comment (adamruzicka) [reply lib/dynflow/executors.rb]: I tried to capture what's going on in the commit message of https://github.com/Dynflow/dynflow/pull/352/commits/7bd35862708e0c181988def1d5910249ab4d7ae5

**Comment by jlsherrill** [comment]
doing a bit more testing with this

**Comment by jlsherrill** [comment]
after cranking down the pool size, i was able to reproduce the original issue and with the fix, i no longer saw dynflow 'hanging' and tasks getting stuck, so +1 from me.  I also wonder if we should increase the foreman db pool size by default in the katello scenario, but thats a discussion for another day

**Comment by adamruzicka** [comment]
@ezr-ondrej you had concerns about this previously. Could you take a look once more?

**Pull request merged**